### PR TITLE
Upgrade to TypeScript

### PR DESCRIPTION
**Pull request by ryanio**
This PR upgrades this repository's src, test, and example files to TypeScript.

It also adds our typical ethereumjs-config setup for linting and building.

### CODE REVIEW

**Comment by ryanio** [comment]
unfortunately it looks like the `index.ts` diff just shows it as a new file, so let me know if there might be an easier way to show the changes between the two (`index.js` => `src/index.ts`).

**Comment by evertonfraga** [comment]
> let me know if there might be an easier way to show the changes between the two (index.js => src/index.ts).

[Not your fault](https://github.com/isaacs/github/issues/900). I can run the `git log --follow` locally while reviewing. Unstoppable @ryanio :)

**Comment by holgerd77** [comment]
Super cool! 😄  👍  Please wait with merging here, I got now the publishing rights from Martin, would like to do a short patch release with the maintenance changes separately before we merge in here.

File: src/index.ts [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+export default class Ethash {
+  dbOpts: Object
+  cacheDB?: LevelUp
+  cache: Buffer[]
+  epoc?: number
+  fullSize?: number
+  cacheSize?: number
+  seed?: Buffer
+
+  constructor(cacheDB?: LevelUp) {
```
Comment (holgerd77) [inline src/index.ts:37]: What is the rationality for adding the optionality of `cacheDB` here and then throw in `loadEpoc` if not present? Wouldn't it make sense to make this mandatory in the constructor then?

File: src/index.ts [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    this.dbOpts = {
+      valueEncoding: 'json'
+    }
+    this.cacheDB = cacheDB
+    this.cache = []
```
Comment (holgerd77) [inline src/index.ts:42]: Whew, a lot better initialization, this was really confusing to have this set to `false` in the JS version.

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+      mix = fnvBuffer(mix, Buffer.concat(newdata))
+    }
+
+    const cmix = Buffer.alloc(mix.length / 4)
+    for (i = 0; i < mix.length / 4; i = i + 4) {
+      const a = fnv(mix.readUInt32LE(i * 4), mix.readUInt32LE((i + 1) * 4))
+      const b = fnv(a, mix.readUInt32LE((i + 2) * 4))
+      const c = fnv(b, mix.readUInt32LE((i + 3) * 4))
+      cmix.writeUInt32LE(c, i)
+    }
```
Comment (holgerd77) [inline src/index.ts:113]: Can't really judge/follow why this has moved into the for loop, can you give a short explanation?

File: src/index.ts [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+      if (!valid) {
+        return cb(valid)
+      }
+
+      async.eachSeries(
+        block.uncleHeaders,
+        (uheader: BlockHeader, cb2: (valid: boolean) => void) => {
+          this._verifyPOW(uheader, (valid3) => {
+            valid = valid3
+            cb2(valid)
```
Comment (holgerd77) [inline src/index.ts:232]: Are you confident on this simplification? 😄  Find this pretty hard to decipher (the old version).

**Comment by evertonfraga** [comment]
> do we actually then want to keep this PR open here and instead @ryanio might do last changes upon the PR on the VM? Would this simplify things?

@holgerd77 Yes, it would simplify things. In fact, the typescript migration is a preparatory step to join the VM monorepo anyways, so it wouldn't add much having this PR merged in, and right after, archive the repo.

File: src/index.ts (hunk 1 above)
Comment (holgerd77) [reply src/index.ts:37]: Ok, thanks.

File: src/index.ts (hunk 3 above)
Comment (holgerd77) [reply src/index.ts:113]: Sorry, misread the brackets from the following js code: 🙂

```javascript
for (i = 0; i < mix.length / 4; i = i + 4) {
    var a = ethHashUtil.fnv(
      mix.readUInt32LE(i * 4),
      mix.readUInt32LE((i + 1) * 4)
    )
    var b = ethHashUtil.fnv(a, mix.readUInt32LE((i + 2) * 4))
    var c = ethHashUtil.fnv(b, mix.readUInt32LE((i + 3) * 4))
    cmix.writeUInt32LE(c, i)
  }
```

So no issue here.

**Pull request closed without merging**